Fixes stack alignment issue in handler_block_trampoline on Win64

ntherning opened the pull request:
On Win64 Mono doesn't (yet) implement fast JIT TLS (mono_get_jit_tls_offset() returns -1). This means that the code in mono_arch_create_handler_block_trampoline() takes the code path in the "else" which is the same code path taken when AOT compiling. The code generated in this code path manipulates the stack pointer to ensure it's aligned properly according to the ABI (RSP-8 is 16-byte aligned on function entry) for the AOT case where the mono_handler_block_guard_trampoline() C function is called directly from the handler_block_trampoline.

But in JIT mode the generated code calls a generated generic trampoline which assumes RSP is 16-byte aligned on function entry. When RSP isn't aligned as expected this generic trampoline will abort the process. Since Win64 Mono always takes this code path the process will be aborted and crash.

This patch splits the "else" into two separate code paths, JIT and AOT. In JIT mode Mono will now generate code similar to when fast JIT TLS is available and the stack pointer will be aligned as expected by the generic trampoline it calls. The AOT code path generates exactly the same code as before.

The tests mono/tests/finally_block_ending_in_dead_bb.exe and mono/tests/block_guard_restore_aligment_on_exit.exe trigger this bug.

vargaz commented:
It would be easier to call a specific trampoline, i.e.:
	guint8 *tramp = mono_create_specific_trampoline (NULL, MONO_TRAMPOLINE_HANDLER_BLOCK_GUARD, mono_get_root_domain (), NULL);

vargaz commented:
Also, it might make sense to remove the 'fastpath', this code doesn't need to be fast.

ntherning commented:
Please elaborate. Should I revert the code and just change the line 

```c
guint8 *tramp = ...
```

in `mono_arch_create_handler_block_trampoline()` to what you suggested?

I'm not sure how the specific trampoline solves the stack alignment issue?

vargaz commented:
The generic trampoline code should not be called directly, it expects to be called from a specific trampoline. So changing that one line should fix the issue imho.

ntherning commented:
Ok, I have done some more debugging and I believe I understand what you mean. It also appears to me that the reading of the TLS is redundant. The generic trampoline will call `mono_handler_block_guard_trampoline()` which reads the return address from the TLS again and returns it to the trampoline which later jumps to that address. I see that there's some unwind info being added in the fast JIT TLS code path. I guess that might be why a proper return address is desired on the call stack? Otherwise one could just push a fake return address, like 0?

vargaz commented:
The tls fast path is simply really needed, its only used on linux anyway.

ntherning commented:
Ok, so I will use a specific trampoline instead and remove the fast path and just call `mono_amd64_handler_block_trampoline_helper()` to get the TLS field, right?

ntherning commented:
Hmmm, looks like `mono_get_root_domain()` returns NULL at this stage. I guess `mono_trampolines_init()` is called too early for the root domain to have been created? Maybe that's the reason why a specific trampoline isn't used in this code?

vargaz commented:
The call to mono_create_handler_block_trampoline () in mono_trampolines_init () can be removed I think, it is called later when needed.

ntherning commented:
Yupp, looks like it can be created on demand instead.

I have made it work with a specific trampoline. The `handler_block_trampoline` must preserve `RSP` before the jump to the specific trampoline otherwise `RSP` will be off when control returns to the managed code. So the `handler_block_trampoline` in JIT mode is now reduced to a single `JMP` which just jumps to the specific trampoline.

Will a call through a specific trampoline mess with unwinding at all? The original code produced a nice call stack in VS. But with the specific trampoline VS has some trouble unwinding and producing a proper call stack. Will this affect unwinding in Mono at runtime and could that be a problem? Maybe this is why the original code "simulated" a CALL?

What makes me a little uncomfortable is the difference in JIT vs AOT code generated for the `handler_block_trampoline`. In AOT mode the `mono_handler_block_guard_trampoline()` C function is called directly from the `handler_block_trampoline`. In JIT mode `handler_block_trampoline` now calls a specific trampoline which calls the `MONO_TRAMPOLINE_HANDLER_BLOCK_GUARD` generic trampoline which calls the `mono_handler_block_guard_trampoline()` C function. Why the difference?

I have pushed an alternative branch (https://github.com/ntherning/mono/commit/6c8065a3c64e8f9080e6faf40fdadf5e7b3a3074) which fixes the issue by unifying the JIT and AOT code paths. The generated code will be almost identical. In this version the JIT code does exactly what was done in AOT mode before. It calls `mono_handler_block_guard_trampoline()` directly from the `handler_block_trampoline`. Please have a look, maybe this is a better approach?

Another thing that's a little odd to me is why the call to `mono_amd64_handler_block_trampoline_helper()` is needed at all? `mono_handler_block_guard_trampoline()` does the same thing, returns the return address from the TLS. Why not just call only this function and jump to the address it returns?

vargaz commented:
This whole code is a mess, will see whenever it can be cleaned up.

ntherning commented:
I understand and agree that a refactoring would be needed here. But that
won't happen anytime soon. We need to eliminate this crash on win64 before
that happens. Are any of my suggestions acceptable to you as a short term
solution? I believe it would good if the same code path is taken on both
JIT and AOT so I would go with that patch. I suggest we go with that
one for now and we add an issue for the refactoring. WDYT?

On Thursday, August 18, 2016, Zoltan Varga <notifications@github.com> wrote:

> This whole code is a mess, will see whenever it can be cleaned up.
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/mono/mono/pull/3412#issuecomment-240678916>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/AAISVU3XNZ7rcN_sawoJ9L5ZCA2Y3He1ks5qhC0igaJpZM4JmhFU>
> .
>

vargaz commented:
Ok.

ntherning commented:
Ok, updated the PR. It looks like I changed a lot in that function but that's mostly because of the changes in indentation from removing the if-block. Below is the diff with whitespaces ignored.

Now, where should I add the task to refactor all this? The runtime Trello card? GitHub issues? Bugzilla?

```diff
diff --git a/mono/mini/tramp-amd64.c b/mono/mini/tramp-amd64.c
index d79b275..cf769a9 100644
--- a/mono/mini/tramp-amd64.c
+++ b/mono/mini/tramp-amd64.c
@@ -767,7 +767,6 @@ mono_amd64_handler_block_trampoline_helper (void)
 gpointer
 mono_arch_create_handler_block_trampoline (MonoTrampInfo **info, gboolean aot)
 {
-	guint8 *tramp = mono_get_trampoline_code (MONO_TRAMPOLINE_HANDLER_BLOCK_GUARD);
 	guint8 *code, *buf;
 	int tramp_size = 64;
 	MonoJumpInfo *ji = NULL;
@@ -778,22 +777,17 @@ mono_arch_create_handler_block_trampoline (MonoTrampInfo **info, gboolean aot)
 	unwind_ops = mono_arch_get_cie_program ();
 
 	/*
-	This trampoline restore the call chain of the handler block then jumps into the code that deals with it.
-	*/
-	if (!aot && mono_get_jit_tls_offset () != -1) {
-		code = mono_amd64_emit_tls_get (code, MONO_AMD64_ARG_REG1, mono_get_jit_tls_offset ());
-		amd64_mov_reg_membase (code, MONO_AMD64_ARG_REG1, MONO_AMD64_ARG_REG1, MONO_STRUCT_OFFSET (MonoJitTlsData, handler_block_return_address), 8);
-		/* Simulate a call */
-		amd64_push_reg (code, AMD64_RAX);
-		mono_add_unwind_op_def_cfa_offset (unwind_ops, code, buf, 16);
-		amd64_jump_code (code, tramp);
-	} else {
-		/*
+	 * This trampoline restore the call chain of the handler block then jumps into the code that deals with it.
 	 * We get here from the ret emitted by CEE_ENDFINALLY.
 	 * The stack is misaligned.
 	 */
-		/* Align the stack before the call */
+	/* Align the stack before the call to mono_amd64_handler_block_trampoline_helper() */
+#ifdef TARGET_WIN32
+	/* Also make room for the "register parameter stack area" as specified by the Windows x64 ABI (4 64-bit registers) */
+	amd64_alu_reg_imm (code, X86_SUB, AMD64_RSP, 8 + 4 * 8);
+#else
 	amd64_alu_reg_imm (code, X86_SUB, AMD64_RSP, 8);
+#endif
 	if (aot) {
 		code = mono_arch_emit_load_aotconst (buf, code, &ji, MONO_PATCH_INFO_JIT_ICALL_ADDR, "mono_amd64_handler_block_trampoline_helper");
 		amd64_call_reg (code, AMD64_R11);
@@ -802,23 +796,33 @@ mono_arch_create_handler_block_trampoline (MonoTrampInfo **info, gboolean aot)
 		amd64_call_reg (code, AMD64_RAX);
 	}
 	/* Undo stack alignment */
+#ifdef TARGET_WIN32
+	amd64_alu_reg_imm (code, X86_ADD, AMD64_RSP, 8 + 4 * 8);
+#else
 	amd64_alu_reg_imm (code, X86_ADD, AMD64_RSP, 8);
+#endif
 	/* Save the result to the stack */
 	amd64_push_reg (code, AMD64_RAX);
+#ifdef TARGET_WIN32
+	/* Make room for the "register parameter stack area" as specified by the Windows x64 ABI (4 64-bit registers) */
+	amd64_alu_reg_imm (code, X86_SUB, AMD64_RSP, 4 * 8);
+#endif
 	if (aot) {
 		char *name = g_strdup_printf ("trampoline_func_%d", MONO_TRAMPOLINE_HANDLER_BLOCK_GUARD);
 		code = mono_arch_emit_load_aotconst (buf, code, &ji, MONO_PATCH_INFO_JIT_ICALL_ADDR, name);
 		amd64_mov_reg_reg (code, AMD64_RAX, AMD64_R11, 8);
 	} else {
-			amd64_mov_reg_imm (code, AMD64_RAX, tramp);
+		amd64_mov_reg_imm (code, AMD64_RAX, mono_get_trampoline_func (MONO_TRAMPOLINE_HANDLER_BLOCK_GUARD));
 	}
 	/* The stack is aligned */
 	amd64_call_reg (code, AMD64_RAX);
+#ifdef TARGET_WIN32
+	amd64_alu_reg_imm (code, X86_ADD, AMD64_RSP, 4 * 8);
+#endif
 	/* Load return address */
 	amd64_pop_reg (code, AMD64_RAX);
 	/* The stack is misaligned, thats what the code we branch to expects */
 	amd64_jump_reg (code, AMD64_RAX);
-	}
 
 	mono_arch_flush_icache (buf, code - buf);
 	mono_profiler_code_buffer_new (buf, code - buf, MONO_PROFILER_CODE_BUFFER_HELPER, NULL);
```

ntherning commented:
Looking at the PR test results this fix doesn't cause any regressions. Can I merge it?

vargaz commented:
Looks ok.
